oauth2/src/test/java/com/linecorp/armeria/internal/common/auth/oauth2/SerialFutureTest.java

```diff
@@ -18,427 +18,263 @@
+        int c = counter.get();
+        try {
+            Thread.sleep(timeout);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
         }
-
-        // invoke test actions in parallel
-        final ForkJoinPool pool = new ForkJoinPool(n);
-        final CompletableFuture<?>[] futures = pool.invokeAll(testActions).parallelStream().map(f -> {
-            try {
-                return f.get();
-            } catch (InterruptedException | ExecutionException e) {
-                throw new RuntimeException(e);
-            }
-        }).toArray(CompletableFuture[]::new);
-
-        // wait for all futures to complete
-        CompletableFuture.allOf(futures).join();
-
-        assertThat(counter).isEqualTo(n);
-        final int[] array = Arrays.stream(futures).mapToInt(future -> (Integer) future.join()).toArray();
-        checkSequentialRange(array, n);
+        c += inc;
+        counter.set(c);
+        return c;
     }
 
-    @Test
-    void testCallAsyncWithExecutor() throws Exception {
-        // execute all actions using dedicated Executor
-        final Executor executor = Executors.newWorkStealingPool();
-        final SerialFuture serialFuture = new SerialFuture(executor);
-
-        final int n = 10;
-        final long timeout = 10L;
-        counter = 0;
+    private static CompletionStage<Integer> timeoutAction(VolatileCounter counter, int inc, long timeout) {
+        return CompletableFuture.completedFuture(simpleTimeoutAction(counter, inc, timeout));
+    }
 
-        final CompletableFuture<?>[] futures = new CompletableFuture<?>[n];
-        // invoke serialFuture#callAsync(Callable) one by one from the test thread
-        // use Executor as part of SerialFuture instance
-        for (int i = 0; i < n; i++) {
-            futures[i] = serialFuture.callAsync(() -> {
-                int c = counter;
-                try {
-                    Thread.sleep(timeout);
-                } catch (InterruptedException e) {
-                    throw new RuntimeException(e);
-                }
-                c += 1;
-                counter = c;
-                return c;
-            }).toCompletableFuture();
+    private static int[] createArithmeticSequence(int start, int inc, int count) {
+        final int[] seq = new int[count];
+        int current = start;
+        for (int i = 0; i < count; i++) {
+            seq[i] = current;
+            current += inc;
         }
-        CompletableFuture.allOf(futures).join();
-
-        assertThat(counter).isEqualTo(n);
-        final int[] array = Arrays.stream(futures).mapToInt(future -> (Integer) future.join()).toArray();
-        checkSequentialRange(array, n);
+        return seq;
     }
 
-    @Test
-    void testCallAsyncWithoutExecutor() throws Exception {
-        final SerialFuture serialFuture = new SerialFuture();
+    private static final class VolatileCounter {
```

Comment:
Could just use `AtomicInteger`?

Reply:
It could, but I did not want to add any synchronization to impact the test. Do you have any particular concern?

Reply:
`AtomicInteger` is just a container of a volatile integer, which basically is identical to `VolatileCounter`, i.e. no need to write a new class for this.

Reply:
Replaced with AtomicInteger